README.md
# ebookdownloader

网文下载器

[![GitHub license](https://img.shields.io/github/license/sndnvaps/ebookdownloader)](https://github.com/sndnvaps/ebookdownloader/blob/master/LICENSE)

[![Build Status](https://travis-ci.org/sndnvaps/ebookdownloader.svg?branch=master)](https://travis-ci.org/sndnvaps/ebookdownloader)[![release_version](https://img.shields.io/github/release/sndnvaps/ebookdownloader.svg)](https://github.com/sndnvaps/ebookdownloader/releases)[![Download Count](https://img.shields.io/github/downloads/sndnvaps/ebookdownloader/total.svg)](https://github.com/sndnvaps/ebookdownloader/releases)

[![Donate on PayPal](https://img.shields.io/badge/support-PayPal-blue?style=flat-square&logo=PayPal)](https://paypal.me/sndnvaps)

[![godoc](https://img.shields.io/badge/godoc-reference-blue.svg)](https://godoc.org/github.com/sndnvaps/ebookdownloader/)

# ebookdl 网文下载器，go 语言版本

## 编译版本依赖条件

1.  go >= 1.16
2.  gcc 必须要安装，用于编译 github.com/andlabs/ui 库

## 安装方法

```bash
go get github.com/sndnvaps/ebookdownloader/cli
go get github.com/sndnvaps/ebookdownloader/qtgui #此项目暂停更新，备份在backup分支当中
go get github.com/sndnvaps/ebookdownloader/http-server #此项目暂停更新，备份在backup分支当中
go get github.com/sndnvaps/ebookdownloader/ui
```

## 使用方法

```bash
.\ebookdownloader.exe --bookid=0_642 --txt #只生成txt文本
.\ebookdownloader.exe --bookid=0_642 --mobi #只生成mobi电子书
.\ebookdownloader.exe --bookid=0_642 --txt --mobi #生成txt 和 mobi
.\ebookdownloader.exe --bookid=0_642 --txt --awz3 #生成txt 和 awz3
.\ebookdownloader.exe --proxy="http://proxyip:proxyport" --bookid=0_642 --mobi #生成mobi电子书，在下载章节的过程中使用 Proxy
.\ebookdownloader.exe --ebhost=xsbiquge.com --bookid=0_642 --txt --mobi #使用xsbiquge.com做为下载源，生成txt 和 mobi
.\ebookdownloader.exe --ebhost=999xs.com --bookid=0_642 --txt --mobi #使用999xs.com做为下载源，生成txt 和 mobi
 .\ebookdownloader.exe --ebhost=999xs.com --bookid=0_642 --txt --mobi --meta #使用999xs.com做为下载源，生成txt,mobi电子书，并生成meta.json文件于小说目录当中
.\ebookdownloader.exe --ebhost=23us.la --bookid=127064 --pv #新功能，用于打印小说的分卷信息，此时不下载小说任何内容
.\ebookdownloader.exe --bookid=0_0642 --json #生成json格式的小说数据
.\ebookdownloader.exe conv --json=".\outputs\我是谁-sndnvaps\我是谁-sndnvaps.json" --txt --mobi #新功能，转换json格式到txt,mobi格式
.\ebookdownloader.exe --help #显示帮助信息
```

## 依赖程序

    1. kindlegen.exe 支持windows平台
    2. kindlegenLinux 支持Linux 平台
    3. kindlegenMac 支持 Mac平台
    4. cli 项目只需要当前目录存在 ./tools/kindlegen（根据不同的平台，名字有所变化）即可运行
       4.1 Linux平台，需要将tools/kindlegenLinux 复制到/usr/bin/目录里面
    5. qemu-i386-static-armhf 支持在linux arm平台上运行 kindlegenLinux
    6. qemu-i386-static-arm64 支持在linux arm64平台上运行 kindlegenLinux
    7. chromedp 项目，用于捉取小说封面
           需要安装chrome浏览器，做支持
    8. ui项目依赖项目
          github.com/andlabs/ui, 用于生成界面
          gtk-3.0，用于构建ui界面

## 懒人模式，直接下载编译好的程序

墙里面使用 gitee

[gitee ebookdownloader release page](https://gitee.com/sndnvaps/ebookdownloader/releases "https://gitee.com/sndnvaps/ebookdownloader/releases")

墙外面使用 github

[github ebookdownloader release page](https://github.com/sndnvaps/ebookdownloader/releases "https://github.com/sndnvaps/ebookdownloader/releases")

---

## 验证下载的文件的完整性（使用 gpg 签名验证)

下载./scripts/gpg_pub_key.gpg 到/home/username/

导入公钥到系统中，以用于验证文件的完整性

```bash
    $gpg --import gpg_pub_key.gpg
```

在 releases 页面下载的文件，将会包含 ebookdownloader*$version*$arch.zip,ebookdownloader_$version\_$arch.zip.sig
验证方法

验证文件完整性

```bash
   $gpg --verify ebookdownloader_v2.0.0_Win_x86_64.zip.sig ebookdownloader_v2.0.0_Win_x86_64.zip
```

得到的信息如下

```bash
F:\gowork\src\github.com\sndnvaps\ebookdownloader>gpg --verify ebookdownloader_v
2.0.0_Win_x86_64.zip.sig ebookdownloader_v2.0.0_Win_x86_64.zip
gpg: Signature made Sun Aug 21 12:11:09 2022
gpg:                using RSA key DA60862463A0D9F2569EF65A309C3FC08F9C9F57
gpg: Good signature from "JimesYang (sndnvaps) <sndnvaps@gmail.com>" [ultimate]
```

## ❤️ 赞助

如果你觉得 EbookDownloader 项目对你有帮助。你可以给我来一杯啤酒！

|                          PayPal 打赏                          |                                                       微信打赏 Pay                                                        |
| :-----------------------------------------------------------: | :-----------------------------------------------------------------------------------------------------------------------: |
| <a href="https://paypal.me/sndnvaps"> paypal.me/sndnvaps </a> | <img style="width:300px;height:300px" src="https://i.loli.net/2021/11/25/6ZLgDa13yrFiNQ2.jpg" alt="Donate_WeChatPay.jpg"> |

---

## 支持的小说网站

|   网站名   |                网址                | 是否支持 |                备注                |
| :--------: | :--------------------------------: | :------: | :--------------------------------: |
|   笔趣阁   | <s><https://www.xsbiquge.com/></s> |    ×     |              网站改名              |
|   笔趣阁   | <s><https://www.vbiquge.com/></s>  |    ×     |              网站改名              |
|   笔趣阁   | <s><https://www.xxbiquge.net/></s> |    ×     |              网站改名              |
|   笔趣阁   | <s><https://www.biqufan.com/></s>  |    ×     |              网站改名              |
|   笔趣阁   |   <s><https://www.biduo.cc/></s>   |    ×     |              网站改名              |
|   笔趣阁   | <s><https://www.biduoxs.com/></s>  |    ×     |              网站改名              |
|   笔趣阁   |      <https://www.biquwu.cc/>      |    ×     |              网站改名              |
|   笔趣阁   |     <http://www.biqugse.com/>      |    √     |
|   笔趣阁   |     <http://www.biqugei.net/>      |    √     |
|   笔趣阁   |      <https://www.zhhbq.com/>      |    √     |
| 西西文学网 |      <http://www.xixiwx.net/>      |    √     |
|  999 小说  |  <s> <https://www.999xs.com/></s>  |    ×     |              网站改名              |
|  999 小说  |  <s> <https://www.899zw.net/></s>  |    ×     |              网站改名              |
|  999 小说  |   <s><https://www.6zw.net/></s>    |    ×     |              网站改名              |
| 顶点小说网 | <s><https://www.booktxt.net/></s>  |    ×     |              网站改名              |
| 顶点小说网 |    <s><https://www.23us.la></s>    |    ×     | 因为网站原因无法打开，暂定无法使用 |

---

## 更新日志

[CHANGELOG](./CHANGELOG "日志文件")

---

## To Do List

     [√]  1. 添加生成封面功能
     [√]  2. 添加不同平台的接口实现
     [√]  3. 添加生成二级目录的方法(已经添加相应的实例)
     [√]  4. 添加界面版本gui
     [√]  5. 添加http-server,做为后端
     [√]  6. 添加linux arm,arm64平台支持
     [√]  7. 需要限制并发数量，因为vps性能有限 -> 目前限制的并发数量为(300+49)*2 = 698
     [√]  8. 使用boltdb记录小说数据（小说下载网站，bookid,uuid->NewV5格式，cover.jpg,mobi,azw3,txt,epub等位置及md5验证信息）
     [√]  9. 添加https://www.biduo.cc/ 小说网站支持
     [√]  10. 添加 https://www.booktxt.net/ 顶点小说网支持
     [√]  11. 添加 https://www.biquwu.cc/ 笔趣阁小说网支持
     [√]  12. 添加处理 作者，书名中包含的特殊字符串的方法
     [√]  13. 使用goqt对界面进行重写
     [ ]  14. 更新顶点小说 23us.la地址为23us.lol ->并重写其爬虫规则
     [√]  15. 打包需要用到的资源到程序中，减少目录数量，kindlegen程序因为技术和安全性考虑，暂不封装到go程序中
     [√]  16. 减少程序运行时的依赖，尝试使用 github.com/andlabs/ui来构建gui界面
     [√]  17. 添加 http://www.biqugei.net/ 小说网站支持
     [√]  18. i添加https://www.zhhbq.com/ 小说网站支持


CHANGELOG
## 更新日志文件

      2022.10.16 go版本
            1. 添加https://www.zhhbq.com/小说网站支持
            2. 更新单元测试ebookdl_test.go,使用 gopkg.in/check.v1 做为测试框架

	2022.8.28 go版本
	       1. windows平台，ui界面，重新生成syso文件，对应用windows平台386,x86_64
		 2. 修改ui界面生成的版本号为v2.0.0.0
		 3. 删除ui/resource.syso以解决在Linux系统的编译连接时问题
		 4. Linux平台，内嵌kindlegen到程序中，索马里时使用go-memexec来加载程序
		 5. Linux平台，实现单个文件可执行，不需要其它依赖文件

	2022.08.26 go版本
	    1. windows平台，使用embed内嵌kindlegen.exe到程序中，运行时使用go-memexec来加载程序
	    2. windows平台，实现单个文件可执行，不需要其它依赖文件

     2022.08.25 go版本
        1. 更新go版本为v1.16+
        2. 采用embed模式来内嵌tpls/fonts文件
        
     2022.08.21 go版本
        1. 删除qtgui版本
        2. 删除http-server
        3. 删除AppImage构建脚本及相关文件

     2022.08.15 go版本
        1. 添加biqugei.net网站支持

     2022.03.27 go版本
        1. Linux 版本，打包生成*.AppImage格式
        2. kindlegenLinux程序需要安装到/usr/bin/目录，不然无法生成mobi格式的电子书

    2022.02.12 go版本
         1. 修改899zw.com地址为6zw.net
         2. ui界面，添加biqugse.com下载支持


    2021.11.24 go版本
         1. 添加biqugse.com网站支持（目前只添加cli命令行支持）

    2021.09.30 go版本
         1. 修改xxbiquge.net地址为biqufan.com(原因网站改网址)
         2. 尝试使用github.com/andlabs/ui库来构建gui界面
         3. 发布测试版本到Release v1.7.8 下面
    
    2021.09.25 go版本
          1. 修改999xs.com地址为899zw.com

    2021.08.20 go版本
          1. 修改biduo.cc地址为biduoxs.com(原因为网站改网址)

    2021.08.19 go版本
	  1. 修改vbiquge.com地址为xxbiquge.net(原因为网站改网址）
    
    2021.02.11 go版本
          1. 修改xsbiquge.com地址为vbiquge.com(原因为网站改网址)
          2. 删除gui版本的文件，彻底删除目录
          3. 修改powershell编译脚本的编译时间定义，改为24小时制

    2021.01.01 go版本
          1. 把tpls/ ,fonts/  目录合并到程序里面，使程序独立性更强
          2.  qtgui版本，添加 检测更新 功能
          3.  尝试使用linuxdeployqt + appimagetool对 qtgui项目进行打包，尝试生成 *.AppImage程序
          4.  生成封面功能，直接由go代码来完成，不再使用tpls/cover.jpg做为封面底版
          5.  qtgui生成windows桌面图标（使用https://github.com/akavel/rsrc项目来实现)
          6.  删除gui版本，功能与qtgui重复

    2020.12.16 go版本
           1. 不再使用小说网站的封面，直接捉取 起点中文网的封面做epub,mobi的封面

    2020.12.14 go版本
           1. 直接使用网站提供的封面，不做生成处理
    
    2020.11.07 go版本
           1. 更新boltdb的测试例子

    2020.05.01 go版本
            1. 对生成的作者和小说名进行检测，去除特殊字符串（感谢 @jinitalk 的错误反馈）
            2. 修复epub无法单独生成问题

    2020.04.04 go版本
            1. 完成boltdb数据库的支持，编码格式使用gob;数据库保存位置为当前目录，命名为ebookdownloader.db
            2. 当插入的uuid数据已经存在，即表示已经有下载过相应的内容；此时采用Update方法对boltdb数据库进行更新
            3. 增加boltdb_test.go的添加内容，添加Update测试

    2020.04.03 go版本
            1. 使用storm/v3库，添加boltdb的功能接口

    2020.03.30 go版本
	        1. 添加西西文学网下载源: http://www.xixiwx.com/
   
    2020.03.17 go版本
             1. 添加笔趣阁下载源：https://www.biquwu.cc/

    2020.03.12 go版本
	          1. 紧急修复gui中的致命错误
			  2. 给windows平台的cli,http-server添加图标(Linux,Mac不支持此方法)

    2020.03.08 go版本
	          1. 顶点小说下载源:booktxt.go 删除下载的txt文件中包含非小说的内容

    2020.03.07 go版本
              1. 修改进度条，使其能完整显示100%进度

    2020.03.06 go版本
              1. 把小说下载源独立成一个package ebookdownloader
              2. 添加小说下载源 booktxt.com
			  3. 更新版本到 1.7.4

    2020.03.04 go版本
               1. meta.json文件添加生成md5信息（对应mobi,azw3,txt,epub小说）
               2. meta.json添加生成uuid信息(对应每本小说，生成规则为UUID_V5)
               3. 添加新下载源 www.biduo.cc
               

    2020.03.03 go版本
               1. Travis-ci:在编译生成程序前，先进行go test测试当前平台是否有问题
			   2. Travis-ci:在编译和打包的时候打印信息
			   3. ebookdownloader:把生成mobi功能，单独成一个文件
			   4. test: go test完成后，删除生成的测试文件和目录

    2020.03.02 go版本
               1. 更新go-epub版本到v1.0.0,感谢@ystyle提交的修改
               2. 添加生成统一isbn功能（当同时生成mobi,azw3,epub的时候才能统一；如果分开生成，isbn码是不一样的）
    
    2020.03.01 go版本
                1. cli版本:添加检查最新版本功能 update_check功能
                2. cli版本：添加生成epub功能
                3. gui版本:添加生成epub功能
                4. http-server版本：添加生成epub功能
                5. gui版本：把字体嵌入到程序当中
                6. 更新版本到 1.7.3

    2020.02.29 go版本
                1. 对代码函数命名进行规范
                2. 更新测试用命ebookdl_test.go文件
                3. 更新下载章节进度条的格式
                4. 把日志文件单独分离出来
                5. 更新PowerShell脚本，添加生成编译版本的功能

      2020.02.18 go版本
                1. 添加PowerShell脚本，用于支持在win10上编译
  
      2020.02.16 go版本
                1. ebookdownloader库：windows平台 kindlegen运行时添加隐藏窗口功能
                2. gui:运行时，隐藏ebookdownloader_cli程序启动时的黑色窗口
                3. 更新版本到1.7.2
                -----
                4. travis-ci: 对生成的程序进行upx -9压缩（Mac平台因为upx有问题，没有包含进去)

      2020.02.14 go版本
                1. http-server:添加创建任务功能和查询任务功能（使用kala Job Schedule)
                2. 对代码进行添加注释
                3. 添加新接口 GetBookBriefInfo只用于获取小说名字，作者，简介信息
                4. gui: 更新代码，修改字体的路径
                5. gui: 现在改用外部调用ebookdownloader_cli来执行下载操作，需要用到配置文件./conf/ebdl_conf.ini

      2020.02.13 go版本
                1. cli: 重新补充完成meta.json信息

      2020.02.12 go版本
                1. 添加gitee项目地址，加快国内的下载速度
                2. 生成txt：修改方法为按一章一章地往磁盘中写入文件，以降低cpu占用率
                3. cli: 添加生成json格式内容，用于把小说保存成json格式，方便线下再进行转换，方便进行二次处理
                4. 添加二级命令conv,用于转换json文件到txt,mobi,azw3格式小说

      2020.02.10 go版本
                1. 添加小说章节分割下载功能(以300章为一个下载单元)
                2. 限制并发数量，目前最大并发数量为 (300+49)*2 = 698
                3. arm版本问题：目前只能生成txt文件，因为生成mobi花费时间过于长(测试平台:Raspberry Pi 3b, cpu过于垃圾和内存太小)
                4. 更新版本到 v1.7.1
                5. 更换字体为: 文泉驿微米黑 Regular

     2020.02.02 go版本
                1. 初步添加kala接口，做为 Job Scheduler
                2. cli版本添加生成meta.json文件功能

     2020.01.28 go版本
                1. ebookdownloader 修改获取章节的规则:替换 <br/> 为 \r\n
                2. http-server 添加鉴权功能，通过/login来获取 token

      2020.01.27 go版本
                1. http-server添加中文件，处理跨域访问问题
                2. 修改小说下载后，保存目录为 ./outputs/小说名-作者/
                3. http-server 添加生成meta.json,用于保存小说作者，小说简介，小说下载网站，小说bookid等信息
                4. http-server 配置文件修改，原来的host定义为外部地址，iner_host定义为内部地址
                
      2020.01.26 go版本
                1. 添加 http-server版本，初始化
                2. 添加qemu-i386-static 支持arm,arm64平台上生成mobi,azw3格式电子书
                3. 更新版本到 v1.7.0

      2020.01.24 go版本
                1. 版本更新到 v1.6.3
                2. 更新到 v1.6.4 用于测试 Travis-ci

ui/form.go
package main

import (
	"context"
	"fmt"
	"strings"
	"time"

	"github.com/andlabs/ui"

	edl "github.com/sndnvaps/ebookdownloader"
	ebook "github.com/sndnvaps/ebookdownloader/ebook-sources"
)

var (
	//Version 版本信息
	Version string = "dirty"
	//Commit git commit信息
	Commit string = "b035048"
	//BuildTime 编译时间
	BuildTime string = "2023-04-22 22:35:21"
)

func makeAboutWindow() ui.Control {
	vbox := ui.NewVerticalBox()
	vbox.SetPadded(true)

	hbox := ui.NewHorizontalBox()
	hbox.SetPadded(true)

	//hbox.Append(vbox, false)

	data1 := fmt.Sprintf("%s\n", "Ebookdownloader")
	data2 := fmt.Sprintf("版本:%s\n", Version)
	data3 := fmt.Sprintf("CommitHash:%s\n", Commit)
	data4 := fmt.Sprintf("编译时间:%s\n", BuildTime)

	announceInfo := "       声明       \n"
	announceInfo += "  本软件用于下载网络上的小说，并生成对应的格式。\n目前支持生成txt,mobi,epub等三种格式。\n"
	announceInfo += "对于本软件在使用过程中造成的问题，一概不负责"
	announceInfoMultiEntry := ui.NewNonWrappingMultilineEntry()

	announceInfoMultiEntry.SetText(announceInfo)
	announceInfoMultiEntry.SetReadOnly(true)

	hbox.Append(ui.NewLabel(data2), false)
	hbox.Append(ui.NewLabel(data3), false)

	vbox.Append(ui.NewLabel(data1), false)
	vbox.Append(hbox, false)
	vbox.Append(ui.NewLabel(data4), false)
	vbox.Append(ui.NewHorizontalSeparator(), false)
	//富文件支持
	vbox.Append(announceInfoMultiEntry, true)

	return vbox
}

// 主页
func makeHomeWindow() ui.Control {
	vbox := ui.NewVerticalBox()
	vbox.SetPadded(true)

	hboxID := ui.NewHorizontalBox() //用于设置BookID
	hboxID.SetPadded(true)

	hboxProxy := ui.NewHorizontalBox() //用于设置Proxy
	hboxProxy.SetPadded(true)

	hboxChooseWebsite := ui.NewHorizontalBox() //用于选择下地小说数据的网站
	hboxChooseWebsite.SetPadded(true)

	hboxSaveType := ui.NewHorizontalBox() //用于设置生成小说的格式
	hboxSaveType.SetPadded(true)

	bookIDLabel := ui.NewLabel("BookID")
	bookIDInputEntry := ui.NewEntry()

	hboxID.Append(bookIDLabel, false)
	hboxID.Append(bookIDInputEntry, false)

	proxyLabel := ui.NewLabel("Proxy")
	proxyInputEntry := ui.NewEntry()

	hboxProxy.Append(proxyLabel, false)
	hboxProxy.Append(proxyInputEntry, false)

	fictionWebsiteCombox := ui.NewCombobox()
	fictionWebsiteCombox.Append("biqugei.net")
	fictionWebsiteCombox.Append("biqugse.com")
	fictionWebsiteCombox.Append("xixiwx.net")
	fictionWebsiteCombox.Append("zhhbq.com")
	fictionWebsiteCombox.SetSelected(0) //设置默认选择为 biqugei.net
	fictionWebsiteLabel := ui.NewLabel("请选择要用到的下载源")

	hboxChooseWebsite.Append(fictionWebsiteCombox, false)
	hboxChooseWebsite.Append(fictionWebsiteLabel, false)

	checkboxTxt := ui.NewCheckbox("txt")
	checkboxMobi := ui.NewCheckbox("mobi")
	checkboxEpub := ui.NewCheckbox("epub")

	hboxSaveType.Append(checkboxTxt, false)
	hboxSaveType.Append(checkboxMobi, false)
	hboxSaveType.Append(checkboxEpub, false)

	group := ui.NewGroup("选择要保存的格式")
	group.SetMargined(true)
	//group.SetChild(ui.NewNonWrappingMultilineEntry())
	group.SetChild(hboxSaveType)

	var bookinfo edl.BookInfo              //初始化变量
	var EBDLInterface edl.EBookDLInterface //初始化接口

	runBtn := ui.NewButton("下载")
	runBtn.OnClicked(func(*ui.Button) {
		bookid := ""
		proxy := ""
		ctx := context.Background()
		ctx, cancel := context.WithTimeout(ctx, time.Second)
		defer cancel()

		if strings.Compare(bookIDInputEntry.Text(), "") != 0 {
			bookid = bookIDInputEntry.Text()
		}

		if strings.Compare(proxyInputEntry.Text(), "") != 0 {
			proxy = proxyInputEntry.Text()
		}
		switch fictionWebsiteCombox.Selected() {
		case 0:
			biqugei := ebook.NewBiqugei()
			EBDLInterface = biqugei //实例化接口
		case 1:
			biqugse := ebook.NewBiqugse()
			EBDLInterface = biqugse //实例化接口
		case 2:
			xixiwx := ebook.NewXixiwx()
			EBDLInterface = xixiwx //实例化接口
		case 3:
			zhhbq := ebook.NewZhhbq()
			EBDLInterface = zhhbq //实例化接口
		}

		bookinfo = EBDLInterface.GetBookInfo(ctx, bookid, proxy)
		bookinfo = EBDLInterface.DownloadChapters(bookinfo, proxy) //下载小说章节内容
		if checkboxTxt.Checked() {                                 //当被选择时，生成txt格式
			bookinfo.SetDownloadCoverMethod(false)
			bookinfo.GenerateTxt()
		}
		if checkboxMobi.Checked() { //当被选择时，生成mobi格式
			bookinfo.SetKindleEbookType(true /* isMobi */, false /* isAzw3 */)
			bookinfo.SetDownloadCoverMethod(true)
			bookinfo.GenerateMobi()
		}
		if checkboxEpub.Checked() { //当被选择时，生成epub格式
			bookinfo.SetDownloadCoverMethod(true)
			bookinfo.GenerateEPUB()
		}

		MsgBoxInfo := fmt.Sprintf("小说名：%s\n作者：%s\n简介：\n\t%s", bookinfo.Name, bookinfo.Author, bookinfo.Description)
		ui.MsgBox(ui.NewWindow("MsgBox", 40, 50, false), "小说已经下载完成", MsgBoxInfo)

	})

	vbox.Append(hboxID, false)
	//分割线
	vbox.Append(ui.NewHorizontalSeparator(), false)
	vbox.Append(hboxProxy, false)
	vbox.Append(ui.NewHorizontalSeparator(), false)
	vbox.Append(hboxChooseWebsite, false)
	vbox.Append(ui.NewHorizontalSeparator(), false)
	//vbox.Append(hboxSaveType, false)
	vbox.Append(group, true)
	vbox.Append(ui.NewHorizontalSeparator(), false)
	vbox.Append(runBtn, false)

	return vbox
}

func InitUI() {
	mainwin := ui.NewWindow("Ebookdownloader", 400, 300, true)
	//关闭
	mainwin.OnClosing(func(*ui.Window) bool {

		ui.Quit()
		return true
	})
	//退出
	ui.OnShouldQuit(func() bool {
		mainwin.Destroy()
		return false
	})

	form := ui.NewForm()
	form.SetPadded(true)

	mainwin.SetChild(form)
	mainwin.SetMargined(true)

	//主菜单
	tab := ui.NewTab()
	mainwin.SetChild(tab)
	mainwin.SetMargined(true)

	tab.Append("主页", makeHomeWindow())
	tab.Append("关于", makeAboutWindow())

	mainwin.Show()
}
